src/bitcoin-miner.cpp

```diff
@@ -0,0 +1,723 @@
+#include <event2/keyvalq_struct.h>
+
+#include <univalue.h>
+
+using namespace std;
+
+// BU add lockstack stuff here for bitcoin-miner, because I need to carefully
+// order it in globals.cpp for bitcoind and bitcoin-qt
+boost::mutex dd_mutex;
+std::map<std::pair<void *, void *>, LockStack> lockorders;
+boost::thread_specific_ptr<LockStack> lockstack;
+
+static const int CONTINUE_EXECUTION = -1;
+
+
+// Internal miner
+//
+// ScanHash scans nonces looking for a hash with at least some zero bits.
+// The nonce is usually preserved between calls, but periodically or if the
+// nonce is 0xffff0000 or above, the block is rebuilt and nNonce starts over at
+// zero.
+bool static ScanHash(const CBlockHeader *pblock, uint32_t &nNonce, uint256 *phash)
+{
+    // Write the first 76 bytes of the block header to a double-SHA256 state.
+    CHash256 hasher;
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << *pblock;
+    assert(ss.size() == 80);
+    hasher.Write((unsigned char *)&ss[0], 76);
+
+    while (true)
+    {
+        nNonce++;
+
+        // Write the last 4 bytes of the block header (the nonce) to a copy of
+        // the double-SHA256 state, and compute the result.
+        CHash256(hasher).Write((unsigned char *)&nNonce, 4).Finalize((unsigned char *)phash);
+
+        // Return the nonce if the hash has at least some zero bits,
+        // caller will check if it has enough to reach the target
+        if (((uint16_t *)phash)[15] == 0)
+            return true;
+
+        // If nothing found after trying for a while, return -1
+        if ((nNonce & 0xfff) == 0)
+            return false;
+    }
+}
+
+
+// Exception thrown on connection error.  This error is used to determine
+// when to wait if -rpcwait is given.
+class CConnectionFailed : public std::runtime_error
+{
+public:
+    explicit inline CConnectionFailed(const std::string &msg) : std::runtime_error(msg) {}
+};
+
+class BitcoinMinerArgs : public AllowedArgs::BitcoinCli
+{
+public:
+    BitcoinMinerArgs(CTweakMap *pTweaks = nullptr)
+    {
+        addHeader(_("Mining options:"))
+            .addArg("blockversion=<n>", ::AllowedArgs::requiredInt,
+                _("Set the block version number. For testing only.  Value must be an integer"))
+            .addArg("cpus=<n>", ::AllowedArgs::requiredInt,
+                _("Number of cpus to use for mining (default: 1).  Value must be an integer"))
+            .addArg("duration=<n>", ::AllowedArgs::requiredInt,
+                _("Number of seconds to mine a particular block candidate (default: 30). Value must be an integer"))
+            .addArg("nblocks=<n>", ::AllowedArgs::requiredInt,
+                _("Number of blocks to mine (default: mine forever / -1). Value must be an integer"));
+    }
+};
+
+
+// This function returns either one of EXIT_ codes when it's expected to stop the process or
+// CONTINUE_EXECUTION when it's expected to continue further.
+static int AppInitRPC(int argc, char *argv[])
+{
```

Comment:
Many of these functions are cut and paste duplicates from bitcoin-cli.cpp.  Seems to make better sense to make them non-static in bitcoin-cli.cpp rather than copying them all to here.

AppInitRPC()
HttpReply()
http_request_done()
CallRpc()